src/user_command/mod.rs

```diff
@@ -0,0 +1,62 @@
+use std::io::{self, BufRead};
+
+use clap::Parser;
+
+pub static VALID_AFFIRMATIVE_RESPONSES: [&str; 2] = ["yes", "y"];
+pub static VALID_NEGATIVE_RESPONSES: [&str; 2] = ["no", "n"];
+static VALID_ACTIONS: [&str; 1] = ["move"];
+
+pub fn get_valid_response(input: String) -> (String, bool) {
+    let response_is_valid = is_valid(&input, &VALID_AFFIRMATIVE_RESPONSES)
+        || is_valid(&input, &VALID_NEGATIVE_RESPONSES);
+    if response_is_valid {
+        return (input, true);
+    }
+    return (
+        format!(
+            "Please provide a valid response: {}, {}.",
+            VALID_AFFIRMATIVE_RESPONSES.join(", "),
+            VALID_NEGATIVE_RESPONSES.join(", ")
+        ),
+        false,
+    );
+}
+
+pub fn get_valid_action(input: String) -> (String, bool) {
+    let action_is_valid = is_valid(&input, &VALID_ACTIONS);
+    if action_is_valid {
+        return (input, true);
+    }
+    return (
+        format!(
+            "Please provide a valid action: {}.",
+            VALID_ACTIONS.join(", ")
+        ),
+        false,
+    );
+}
+
+pub fn get_user_input(f: &dyn Fn(&String) -> (String, bool)) -> String {
+    let stdin = io::stdin();
+
+    let mut buffer = String::with_capacity(2048);
+    // Lock our standard input to eliminate synchronization overhead (unlocks when dropped)
+    let mut stdin = io::stdin().lock();
```

Comment:
I'm curious about this, and maybe I just haven't learned enough about Rust yet, but -

line 40 defines an immutable variable, right? but then line 44 defines the same variable name, but mutable, so is this still a "shadow" variable since the first one was supposed to be immutable?